write-fonts/src/font_builder.rs

```diff
@@ -37,16 +48,57 @@ impl TableDirectory {
 }
 
 impl<'a> FontBuilder<'a> {
-    pub fn add_table(&mut self, tag: Tag, data: impl Into<Cow<'a, [u8]>>) -> &mut Self {
+    /// Create a new builder to compile a binary font
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Add a table to the builder.
+    ///
+    /// The table can be any top-level table defined in this crate. This function
+    /// will attempt to compile the table and then add it to the builder if
+    /// successful, returning an error otherwise.
+    pub fn add_table<T>(&mut self, table: &T) -> Result<&mut Self, BuilderError>
+    where
+        T: FontWrite + Validate + TopLevelTable,
+    {
+        let tag = T::TAG;
+        let bytes = crate::dump_table(table).map_err(|inner| BuilderError { inner, tag })?;
+        Ok(self.add_raw(tag, bytes))
+    }
+
+    /// A builder method to add raw data for the provided tag
+    pub fn add_raw(&mut self, tag: Tag, data: impl Into<Cow<'a, [u8]>>) -> &mut Self {
         self.tables.insert(tag, data.into());
         self
     }
 
+    /// Copy each table from the source font if it does not already exist
+    pub fn copy_missing_tables(&mut self, font: FontRef<'a>) -> &mut Self {
+        for record in font.table_directory.table_records() {
+            let tag = record.tag();
+            if !self.tables.contains_key(&tag) {
+                let data = match font.data_for_tag(tag) {
+                    Some(data) => data,
+                    None => {
+                        log::warn!("data for '{tag}' is malformed");
+                        Default::default()
+                    }
+                };
```

Comment:
My preference would be to skip the table here (or return an error) rather than adding an empty one.